File: react-kanban/src/types/drag_and_drop.ts
import type { ColumnType } from "./serialize";
import type { TaskId } from "./state";

export const DRAGGABLE_ITEM_MIME_TYPE = "application/x-kanban-task+json";
export const DRAGGABLE_ITEM_TEXT_FALLBACK = "text/plain";

export type DraggableItemDto = {
  taskId: TaskId;
  sourceColumnType: ColumnType;
};

export function createDraggableItemDto(
  taskId: TaskId,
  sourceColumnType: ColumnType,
): DraggableItemDto {
  return {
    taskId,
    sourceColumnType,
  };
}

export function draggableItemFromPayload(
  payload: string,
): DraggableItemDto | null {
  try {
    const parsed = JSON.parse(payload) as Partial<DraggableItemDto>;

    if (!parsed.taskId || !parsed.sourceColumnType) {
      return null;
    }

    return {
      taskId: parsed.taskId,
      sourceColumnType: parsed.sourceColumnType,
    };
  } catch {
    return null;
  }
}

export function draggableItemToPayload(data: DraggableItemDto): string | null {
  try {
    return JSON.stringify(data);
  } catch {
    return null;
  }
}

File: leptos-kanban/src/types/drag_and_drop.rs
use serde::{Deserialize, Serialize};

use crate::types::serialize::{ColumnType, TaskId};

pub const DRAGGABLE_ITEM_MIME_TYPE: &str = "application/x-leptos-kanban-task+json";
pub const DRAGGABLE_ITEM_TEXT_FALLBACK: &str = "text/plain";

#[derive(Clone, Copy, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct DraggableItemDto {
    pub task_id: TaskId,
    pub source_column_type: ColumnType,
}

impl DraggableItemDto {
    pub fn new(task_id: TaskId, source_column_type: ColumnType) -> Self {
        Self {
            task_id,
            source_column_type,
        }
    }

    pub fn from_payload(payload: &str) -> Option<Self> {
        serde_json::from_str(payload).ok()
    }

    pub fn to_payload(self) -> Option<String> {
        serde_json::to_string(&self).ok()
    }
}
